live-ingester/src/main/java/org/alfresco/hxi_connector/live_ingester/adapters/storage/PreSignedUrlRequester.java

```diff
@@ -0,0 +1,146 @@
+
+import org.alfresco.hxi_connector.live_ingester.adapters.config.HxInsightApiConfig;
+import org.alfresco.hxi_connector.live_ingester.domain.exception.LiveIngesterRuntimeException;
+import org.alfresco.hxi_connector.live_ingester.domain.ports.storage.StorageLocationRequest;
+import org.alfresco.hxi_connector.live_ingester.domain.ports.storage.StorageLocationRequester;
+
+@Component
+@Slf4j
+public class PreSignedUrlRequester extends RouteBuilder implements StorageLocationRequester
+{
+    private static final String LOCAL_ENDPOINT = "direct:" + PreSignedUrlRequester.class.getSimpleName();
+    static final String STORAGE_LOCATION_PROPERTY = "preSignedUrl";
+    private static final String CONTENT_TYPE_PROPERTY = "contentType";
+    private static final String NODE_ID_PROPERTY = "objectId";
+    private static final int EXPECTED_STATUS_CODE = 201;
+
+    private final CamelContext camelContext;
+    private final ObjectMapper objectMapper;
+
+    private final String url;
+    private final String username;
+    private final String password;
+
+    @Autowired
+    public PreSignedUrlRequester(CamelContext camelContext, ObjectMapper objectMapper, HxInsightApiConfig.Properties config)
+    {
+        super(camelContext);
+        this.camelContext = camelContext;
+        this.objectMapper = objectMapper;
+        this.url = config.url().storageLocationRequest();
+        this.username = config.username();
+        this.password = config.password();
+    }
+
+    @Override
+    public void configure()
+    {
+        from(LOCAL_ENDPOINT)
+                .marshal()
+                .json()
+                .to(url + "?httpMethod=POST" +
+                        "&authMethod=Basic" +
+                        "&authUsername=" + username +
+                        "&authPassword=" + password +
+                        "&authenticationPreemptive=true");
+    }
+
+    @Override
+    public URL requestStorageLocation(StorageLocationRequest storageLocationRequest)
+    {
+        Map<String, String> request = Map.of(NODE_ID_PROPERTY, storageLocationRequest.nodeId(),
+                CONTENT_TYPE_PROPERTY, storageLocationRequest.contentType());
+
+        Message message = camelContext.createProducerTemplate()
+                .send(LOCAL_ENDPOINT, exchange -> exchange.getIn().setBody(request))
+                .getMessage();
+
+        return extractUrlFromMessage(message);
+    }
+
+    private URL extractUrlFromMessage(Message message)
+    {
+        String responseBody = message.getBody(String.class);
+        int statusCode = message.getHeader(Exchange.HTTP_RESPONSE_CODE, Integer.class);
+        verifyStatusCode(statusCode, responseBody);
+
+        return extractUrlFromResponseBody(responseBody);
+    }
+
+    private void verifyStatusCode(int actual, String responseBody)
+    {
+        if (actual != EXPECTED_STATUS_CODE)
+        {
+            log.error("Unexpected response body: {}", responseBody);
```

Comment:
I guess this will change in the future anyway as we handle all the error cases, but the log message is misleading since the header is unexpected, not the body.

Reply:
Unless you mean `"Unexpected response. Body: {}"` :slightly_smiling_face:

Reply:
Yes, second option - I was rather thinking about: `"Unexpected response. Body: {}"`. Fixed.